src/vmm/src/lib.rs

```diff
@@ -461,6 +467,143 @@ impl Vmm {
     }
+
+    /// Creates a snapshot of a paused microVM.
+    #[cfg(target_arch = "x86_64")]
+    pub fn create_snapshot(
+        &mut self,
+        params: CreateSnapshotParams,
+    ) -> std::result::Result<(), CreateSnapshotError> {
+        let vcpu_states = self.save_vcpu_states()?;
+
+        let vm_state = self
+            .vm
+            .save_state()
+            .map_err(CreateSnapshotError::SaveVmState)?;
+
+        let device_states = self.save_mmio_device_states();
+
+        let mem_size_mib =
+            self.guest_memory()
+                .map_and_fold(0, |(_, region)| region.len(), |a, b| a + b)
+                >> 20;
+
+        let microvm_state = MicrovmState {
+            vm_info: VmInfo { mem_size_mib },
+            vm_state,
+            vcpu_states,
+            device_states,
+        };
+
+        self.snapshot_memory_to_file(&params.mem_file_path, mem_size_mib, params.snapshot_type)?;
+
+        let mut snapshot_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(params.snapshot_path)
+            .map_err(CreateSnapshotError::SnapshotBackingFile)?;
+
+        // TODO: fallback to current version.
+        let version_map = VersionMap::new();
+        let version = match params.version {
+            Some(version) => version,
+            _ => 1,
+        };
+        let mut snapshot = Snapshot::new(version_map.clone(), version);
+        snapshot
+            .save(&mut snapshot_file, &microvm_state)
+            .map_err(CreateSnapshotError::SerializeMicrovmState)?;
+
+        Ok(())
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn save_vcpu_states(&mut self) -> std::result::Result<Vec<VcpuState>, CreateSnapshotError> {
+        for handle in self.vcpus_handles.iter() {
+            handle
+                .send_event(VcpuEvent::SaveState)
+                .map_err(CreateSnapshotError::SignalVcpu)?;
+        }
+
+        let vcpu_responses = self
+            .vcpus_handles
+            .iter()
+            // `Iterator::collect` can transform a `Vec<Result>` into a `Result<Vec>`.
+            .map(|handle| {
+                handle
+                    .response_receiver()
+                    .recv_timeout(Duration::from_millis(400))
+            })
+            .collect::<std::result::Result<Vec<VcpuResponse>, RecvTimeoutError>>()
+            .map_err(|_| CreateSnapshotError::SaveVcpuState(None))?;
+
+        let vcpu_states = vcpu_responses
+            .into_iter()
+            .map(|response| match response {
+                VcpuResponse::SaveState(state) => Ok(*state),
+                VcpuResponse::SaveStateFailed(err) => {
+                    Err(CreateSnapshotError::SaveVcpuState(Some(err)))
+                }
+                _ => Err(CreateSnapshotError::SaveVcpuState(None)),
+            })
+            .collect::<std::result::Result<Vec<VcpuState>, CreateSnapshotError>>()?;
+
+        Ok(vcpu_states)
+    }
+
+    fn snapshot_memory_to_file(
```

Comment:
I think we can simplify by defining the `trait DumpMemory`:
```
trait DumpMemory<T> {
     // No need for a mem size parameter, it can be computed.
     fn dump<T: std::io::Write>(writer: &mut T) -> std::io::Result;
     fn dump_diff<T: std::io::Write>(writer: &mut T) -> std::io::Result; // might need a different Result type since this calls into KVM.
}
```
Next implement this for GuestMemory. Then you can just call self.guest_memory().dump() or dump_diff().

Reply:
done, I implemented this interface in `memory_dump.rs`